feat: Add support for library instrumentation

arbrown opened the pull request.

arbrown committed:
Add .python-version to .gitignore

arbrown committed:
Add initial class/test for instrumentation_source

arbrown committed:
Add version and truncate logic

arbrown committed:
Add instrumentation tests and severity info

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Add method to update and validate existing info

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Add .python-version to gitignore

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Implement hook to add instrumentation for logger

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Add tests for logger instrumentation logic

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Update structured log handler to emit info

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Refactor structured log and add unit test

arbrown committed:
Merge branch 'main' into instrumentation

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

arbrown committed:
Add side effect to unit test

* Ensure that instrumentation info is only called once

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

release-please[bot] committed:
chore(main): release 3.1.0 (#479)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

arbrown committed:
docs: Change button in README to .png file (#554)

* fix: Change button to .png file

* Change the "Guide Me" button to a .png file to work with RST `image`
* Avoids `raw` directive disabled by PyPi
* Fixes #553

* Fix unexpected unindent in docs

Add copy of png file to render correctly in GitHub
readme and Sphinx generated docs.

arbrown committed:
Update to environment submodule

arbrown committed:
Fix linter errors

release-please[bot] committed:
chore(main): release 3.1.1 (#557)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

arbrown committed:
Update env-tests submodule

arbrown committed:
Merge branch 'main' into instrumentation

arbrown committed:
Minor format update

arbrown committed:
Fix system test to skip diagnostic log entry

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

losalex reviewed (approved):
Great job, Drew! Please take a look on some comments I left, but overall looks good!

daniel-sanche commented on google/cloud/logging_v2/handlers/structured_log.py:
```diff
@@ -16,10 +16,13 @@
 """
 import collections
 import json
+import logging
```
When using StructuredLogHandler, the log passed through Python's logging std lib before reaching our library. Would that be worth capturing as an instrumentation source? Do we want to be able to differentiate between different handlers?

arbrown replied:
I think the spec would allow for that, but that might be overkill for the first version of instrumentation source.  I think just knowing overall library usage and version numbers is the primary goal.  But I'll defer to @losalex on this.

daniel-sanche replied:
FWIW I noticed nodejs seems to be [capturing both node and the bunyan integration library info](https://github.com/googleapis/nodejs-logging-bunyan/blob/101822bc4ec43d579287bdf02846f5b62803f76a/system-test/logging-bunyan.ts#L86)

But yeah, we can probably decide if we want to increase the resolution of the instrumentation later. It may be worth thinking about the naming choices for the different layers of instrumentation should be named ahead of time though, because that could be hard to change later. But I think using `python` should probably work for the base entry

losalex replied:
We should capture all libraries we own.

arbrown replied:
I think for now we probably don't need to capture std lib logging then.  It might be useful later, but for now the focus is on the layers we control.

daniel-sanche replied:
I do think the integration layer for the stdlib is kind of a second library on its own, even though it's in the same repo. And I think it would be very useful to see how many people are using the integration vs using the API directly

But yeah agreed, that can easily be added later

losalex replied:
I would say we should look into it if you think it is important - I probably should mention before that libraries we own is a must, but others are definitely subjects to consider. Lets sync on this

arbrown committed:
Merge branch 'main' into instrumentation

arbrown committed:
Update truncation logic based on feedback

arbrown committed:
Update environment tests

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

arbrown committed:
Fix broken unit test

arbrown committed:
Fix broken unit test

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+        A string containing either 'str' or a truncated version of
+        'str' with an asterisk at the end
+    """
+    if len(str) > max_length:
```
Fixed based on feedback.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown commented on google/cloud/logging_v2/handlers/structured_log.py:
```diff
@@ -84,3 +87,13 @@ def format(self, record):
+    def emit(self, record):
+        if google.cloud.logging_v2.instrumentation_emitted is False:
+            self.emit_instrumentation_info()
+        super().emit(record)
```
The main thing I can think of is that we don't necessarily want to emit a diagnostic log if no other logs are emitted.  That might not be worth worrying about though, what do you think?

daniel-sanche replied:
I guess I'll leave that to @losalex. Sending a diagnostic every time a new client is initialized up sounds reasonable to me, but it sounds like that's not how the spec was designed? We'd want to make sure we're consistent at least

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+       obj: JSON object with library information
+    """
+    name = name if name is not None else _PYTHON_LIBRARY_NAME
+    version = version if version is not None else _LIBRARY_VERSION
```
No good reason.  They won't ever change, so I think putting them in the argument defaults is fine.  I'll change it.

arbrown committed:
Change default name/version

arbrown committed:
Refactor add_instrumentation

Return a new list instead of a mutated original
Do not return after first log without info

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+            is_written = True
+        else:
+            diagnostic_entry = create_diagnostic_entry(**kw)
+            entries.insert(0, diagnostic_entry.to_api_repr())
```
Good points.  I've refactored to return a new list and avoid inserts during iteration.

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+        else:
+            diagnostic_entry = create_diagnostic_entry(**kw)
+            entries.insert(0, diagnostic_entry.to_api_repr())
+        return entries
```
Fixed.

arbrown committed:
Add more documentation to validation methods

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+
+
+def add_instrumentation(entries, **kw):
+    """Validate or add instrumentation information to a list of entries
```
Done.

arbrown committed:
Refactor add_instrumentation to be more pythonic

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+            and hasattr(entry, "payload")
+            and entry.payload is dict
+            and _DIAGNOSTIC_INFO_KEY in entry.payload
+            and _INSTRUMENTATION_SOURCE_KEY in entry.payload[_DIAGNOSTIC_INFO_KEY]
```
Ok.  I refactored it to break out of the loop once it had found a valid entry.  It also catches exceptions for entries that do not have instrumentation info.  There is nothing to handle there, so I just `pass`.  PTAL.

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+        if (
+            is_written is False
+            and hasattr(entry, "payload")
+            and entry.payload is dict
```
Replaced this section with exception handling logic if the key/attr are not found.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+
+
+def validate_and_update_instrumentation(existing_info=None):
+    """Validate existing instrumentation info and append the final entry
```
It only matters if we ever expect to receive a log entry with existing (or partial) instrumentation source.  This is probably not going to happen with the library as it stands, but in the future, something could be built on top of this library (like the bunyan or winston integrations for nodejs) that adds a valid 'layer' of instrumentation info that we don't want to clobber.

It would certainly simplify things to remove this logic and just always emit a single entry with just this library's info... What do you think?

arbrown commented on google/cloud/logging_v2/logger.py:
```diff
@@ -155,7 +157,13 @@ def _do_log(self, client, _entry_class, payload=None, **kw):
+        entries = [api_repr]
+        if google.cloud.logging_v2.instrumentation_emitted is False:
+            partial_success = True
+            entries = add_instrumentation(entries, **kw)
```
Is there any harm in adding `partial_success=True` to a call to `write_entries` with a single entry?  Because that's the current behavior for log entries if the first log entry already had instrumentation info.

losalex replied:
Hm... When you add a record with instrumentation data, it means there are at least 2 records.

arbrown committed:
Update environemnt tests

arbrown committed:
Merge branch 'main' into instrumentation

arbrown committed:
Refactor _is_valid and add test

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+        'str' with an asterisk at the end
+    """
+    if len(str) > max_length:
+        return str[: max_length - 1] + "*"
```
Yes

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+                break
+            if _is_valid(info):
+                info_stack.append(
+                    _get_instrumentation_source(info["name"], info["version"])
```
I simplified the function a bit, but I'd like to leave it in place and not raise an error.  The intention here was to 'clean up' potential diagnostic info added manually by users so we don't have stacks of hundreds of layers of bogus info.  For that reason, we verify any info that is already there, and just take the last 3 valid ones (if any exist.) Do you think that explicitly raising an error is a better solution here?

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+            _INSTRUMENTATION_SOURCE_KEY: [_get_instrumentation_source(name, version)]
+        }
+    }
+    kw["severity"] = "INFO"
```
The spec indicates that it is for this entry

arbrown committed:
Add more detail to method documentation

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,156 @@
+
+
+def create_diagnostic_entry(name=None, version=None, **kw):
+    """Create a diagnostic log entry describing this library
```
Done

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Move methods to private

arbrown commented on google/cloud/logging_v2/handlers/structured_log.py:
```diff
@@ -84,3 +87,13 @@ def format(self, record):
         return gcp_payload
+
+    def emit(self, record):
+        if google.cloud.logging_v2.instrumentation_emitted is False:
```
I think I prefer to leave it explicit.  It seems a bit clearer than relying on an opaque return value, though I'd be interested in @daniel-sanche 's thoughts on which method seems more pythonic.

losalex replied:
The problem here might be potential race condition (despite a fact that Python usually used under GIL) - hence you want to minimize time between variable value is taken and set... Given a fact that this library can be used anywhere, it might come to a problem of multiple records written since variable is checked in one function and set in another.
Goof programming style in such case would be calling atomic "get and set" function, but apparently I am not sure one exists in Python - in this case you could make that "get and set" in `setInstrumentationEmittedState()` at least to keep good programming style

daniel-sanche replied:
Yeah I think in general the Python style prefers not to create boilerplate setters and getters when not necessary, but Alex makes a good point that there could be a race condition here, so we should make this atomic

arbrown replied:
Ok.  I think I want to work on that name a bit. `setInstrumentationEmittedState()` implies that it's just setting the variable (presumably to true, with or without an explicit argument.) I'm not sure how best to capture that we're returning the previous state other than very clear documentation.  Maybe it's not important, but if we could make it clear in the name, that would be ideal.

losalex replied:
Yeah, make sure that it is documented at least

daniel-sanche commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,173 @@
+    """
+    return {
+        "name": _truncate_string(name, _MAX_NAME_LENGTH),
+        "version": _truncate_string(version, _MAX_VERSION_LENGTH),
```
suggestion: since the truncation logic can be written as a simple one-liner, and since it's used in one place, and since there's nothing else going on in this function, I'd probably suggest just doing the truncation in-place here:

```
def _get_instrumentation_source(name=_PYTHON_LIBRARY_NAME, version=_LIBRARY_VERSION):
    source = {"name": name, "version": version}
   # truncate strings to no more than _MAX_NAME_LENGTH characters
   for key, val in source.items():
     source[key] = val if len(val) <= _MAX_NAME_LENGTH else f"{val[:_MAX_NAME_LENGTH]}*"
```

But if you think it'll be useful in other places later, it's probably better to keep it as a separate function

arbrown replied:
That looks great.  Thanks for the suggestion.

daniel-sanche commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,173 @@
+
+
+def _is_valid(info):
+    """Validates an existing instrumentation_source entry
```
is it valid iff the name start with `python`? (This should probably be included in the comments if you keep this function)

arbrown replied:
Done (future validations may be more strict)

daniel-sanche reviewed (commented):
I added a couple more small comments. I have the feeling it could be refactored and simplified some more if you have the time, but overall looks good!

arbrown committed:
Change instumentation_added to private

arbrown commented on tests/unit/test__instrumentation.py:
```diff
@@ -0,0 +1,100 @@
+        self.assertEqual(expected_name, self._get_diagonstic_value(entry, "name"))
+        self.assertEqual(expected_version, self._get_diagonstic_value(entry, "version"))
+
+    def test_validate_and_update_instrumentation_adds_info(self):
```
Added

arbrown committed:
Fix some issues with validation method

Add a test

arbrown committed:
Fix bug in _add_instrumentation

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,173 @@
+    new_entries = []
+    for entry in entries:
+        if is_written:
+            break
```
Yes, I just removed this check.  I think it was a leftover from a previous approach.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,173 @@
+        if is_written:
+            break
+        try:
+            current_info = entry.payload[_DIAGNOSTIC_INFO_KEY][
```
We only ever enter this method (and check) if instrumentation info has not been emitted.  Thus, we do not even inspect/check entries after the first call to _do_log.

arbrown committed:
Simplify string truncation

arbrown committed:
Merge branch 'main' into instrumentation

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

arbrown commented on google/cloud/logging_v2/_instrumentation.py:
```diff
@@ -0,0 +1,161 @@
+    """
+    is_written = False
+    new_entries = []
+    for entry in entries:
```
Simplify by assuming we need to emit exactly once.

arbrown replied:
On second thought, leave in for future expansion with different layers from handlers, etc...

daniel-sanche committed:
Merge branch 'main' into instrumentation

arbrown committed:
Remove unused import to fix lint

arbrown committed:
Merge branch 'main' into instrumentation

arbrown committed:
Remove validate_and_update_instrumentation

Simplify code by adding a single instrumentation entry
instead of validating potential existing entries

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arbrown committed:
Remove _is_valid code (no longer checked)

arbrown committed:
Run nox blacken

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'instrumentation' of https://github.com/googleapis/python-logging into instrumentation

arbrown committed:
Remove extraneous unit test
